File: api/routes/news.go
package routes

import (
	"context"
	"strings"

	"github.com/dickeyy/tip-dev/api/services"
	"github.com/dickeyy/tip-dev/api/utils"
	"github.com/gofiber/fiber/v2"
)

// enter email into newsletter
func EnterNewsletter(c *fiber.Ctx) error {
	type request struct {
		EmailAddress string `json:"email_address"`
	}

	input := new(request)
	if err := c.BodyParser(input); err != nil {
		return c.SendStatus(400)
	}

	invalid := utils.Validate(input)
	if invalid == "EmailAddress" {
		return c.Status(400).JSON(&fiber.Map{
			"message": "Invalid email address",
		})
	}

	input.EmailAddress = strings.TrimSpace(input.EmailAddress)
	input.EmailAddress = strings.ToLower(input.EmailAddress)

	res := services.Redis.SAdd(context.Background(), "tdev-newsletter", input.EmailAddress).Val()
	if res == 0 {
		return c.Status(400).JSON(&fiber.Map{
			"message": "Email already exists",
		})
	}

	return c.Status(200).JSON(&fiber.Map{
		"message": "success",
	})
}

// get all emails from newsletter
func GetNewsletter(c *fiber.Ctx) error {
	// get all emails from redis
	emails := services.Redis.SMembers(context.Background(), "tdev-newsletter").Val()

	return c.Status(200).JSON(&fiber.Map{
		"emails": emails,
	})
}

// delete an email from newsletter
func DeleteNewsletter(c *fiber.Ctx) error {
	type request struct {
		EmailAddress string `json:"email_address"`
	}

	input := new(request)
	if err := c.BodyParser(input); err != nil {
		return c.SendStatus(400)
	}

	invalid := utils.Validate(input)
	if invalid == "EmailAddress" {
		return c.Status(400).JSON(&fiber.Map{
			"message": "Invalid email address",
		})
	}

	input.EmailAddress = strings.TrimSpace(input.EmailAddress)
	input.EmailAddress = strings.ToLower(input.EmailAddress)

	res := services.Redis.SRem(context.Background(), "tdev-newsletter", input.EmailAddress).Val()
	if res == 0 {
		return c.Status(400).JSON(&fiber.Map{
			"message": "Email does not exist",
		})
	}

	return c.Status(200).JSON(&fiber.Map{
		"message": "success",
	})
}

File: js-api/routes/news.js
import express from 'express';
const router = express.Router();
import client from '../utils/db.js';
import dotenv from 'dotenv';
import { redis } from '../utils/redis.js';

dotenv.config();

const db = client.db(process.env.DATABASE_NAME);

// enter newsletter, unprotected
router.post("/enter", (req, res) => {
    const { email } = req.body;

    // make sure the email is valid
    if (!email) {
        return res.status(400).json({ error: { message: "missing email" } });
    }

    // make sure the format is valid
    const emailRegex = /\S+@\S+\.\S+/;
    if (!emailRegex.test(email)) {
        return res.status(400).json({ 
            error: {
                message: "invalid email format"
            }
        });
    }

    // check if email is already in redis
    redis.sIsMember('td-newsletter', email).then((result) => {
        if (result) {
            return res.status(400).json({ 
                error: {
                    message: "email already exists"
                }
            });
        } else {
            // add email to redis
            redis.sAdd('td-newsletter', email).then((result) => {
                return res.json({ message: "success" });
            })
        }
    })
})

// get all newsletter members, unprotected
router.get("/members/get/all", async (req, res) => {
    let members = await redis.sMembers("td-newsletter");

    return res.json({ members: members });
})

export { router as newsRouter }